File: math/primenumbers/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let max=0
    let m=prices[0]

    for(let i=1;i<prices.length;i++){
        if(m>prices[i]){
             m=prices[i]
        }
        if(prices[i-1]<prices[i]){
           
            let sum=prices[i]-m
            if(sum>max){
                max=sum
            }
        }
    }

    return max
    
};

File: math/primenumbers/121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int max=Integer.MIN_VALUE;
        int maxprofit=0;
        int n=prices.length;
        int[] suffix=new int[n];


        for(int i=n-1;i>=0;i--){
            max=Math.max(max,prices[i]);
            suffix[i]=max;
        }

        for(int i=0;i<n;i++){
            int diff=suffix[i]-prices[i];
            if(diff>maxprofit){
                maxprofit=diff;
            }
        }

        return maxprofit;
    }
}